### ISSUE TITLE
Feature Request: Upgrade Packages

### ISSUE THREAD

**Issue by AlJohri** [Homebrew/homebrew-bundle#37]
It would be nice to have the ability to upgrade packages. This could either be automatically (if an upgrade exists for a package in the Brewfile, then upgrade it) or a separate dsl so its a tad more explicit (such as 'brew_upgrade').

This is inspired by thoughtbot's laptop setup script: https://github.com/thoughtbot/laptop/blob/master/mac#L56

```bash
brew_install_or_upgrade() {
  if brew_is_installed "$1"; then
    if brew_is_upgradable "$1"; then
      fancy_echo "Upgrading %s ..." "$1"
      brew upgrade "$@"
    else
      fancy_echo "Already using the latest version of %s. Skipping ..." "$1"
    fi
  else
    fancy_echo "Installing %s ..." "$1"
    brew install "$@"
  fi
}
```

I'd love to submit a PR for this feature, just interested to hear which method is more likely to be accepted.

**Comment by jamesaanderson** [comment]
How about something like `brewdle update`?

**Comment by AlJohri** [comment]
That sounds like a good solution. Would ```brewdle update``` install packages if they aren't already installed or just upgrade the existing packages?

I can see how ```brewdle update``` is more inline with Bundler but saying ```brewdle upgrade``` may be more inline with Homebrew. Not sure. Thoughts?

**Comment by jamesaanderson** [comment]
I think it would be fine to install packages as well. As for the command, I'd alias `brewdle update` and `brewdle upgrade`

**Comment by AlJohri** [comment]
Cool. I haven't written tests yet but does this look alright?

https://github.com/AlJohri/homebrew-brewdler/commit/5ab122279aeea3397ac510532ea67de86cb774a1

**Comment by jamesaanderson** [comment]
Looks good though not sure why you added `"install"` to `dsl_spec.rb` and not feeling how you're passing the mode into `initialize`.

**Comment by AlJohri** [comment]
I was iffy on that too. I needed the ```brew``` method in the dsl to change depending on the ```@mode```. I changed ```dsl_spec.rb``` to match my changes to ```initialize``` (gets 2 arguments). I'll try to think of something else - let me know if you have a better way.

**Comment by jamesaanderson** [comment]
Regarding `dsl_spec.rb` pass in install or upgrade based on the command run but the syntax of the `Brewfile` shouldn't change.

**Comment by AlJohri** [comment]
It doesn't change. The current solution is:

```ruby
Brewdler::Dsl.new(brewfile, "install").process
```

**Comment by jamesaanderson** [comment]
Got it. Good work! Just add some tests and try to come up with a better solution to passing in install or upgrade (maybe create an install or upgrade method?)

**Comment by mikemcquaid** [comment]
I'd be tempted to have `Brewfile` upgrade by default on reruns given that compared to e.g. RubyGems we don't really allow a way to install and pin an older version to allow a team to share versions. As a result it seems better to have everyone on the latest version than requiring an explicit upgrade command.

@AlJohri Could you may a PR; makes review a bit easier? Thanks!

**Comment by jamesaanderson** [comment]
@mikemcquaid The only thing that worries me about auto updating is that an update could be incompatible with your machine or config

**Comment by mikemcquaid** [comment]
@jamesaanderson It won't remove the old version without intervention and it won't unlink it until the new install is successfully installed and linked.

**Comment by andrew** [comment]
I'd agree, going with "install or upgrade to latest" by default would be good to go with, perhaps with a cli flag for the more caution option, something like `--no-upgrade`?

**Comment by AlJohri** [comment]
I like that idea. I can implement the ```--no-upgrade``` if everyone agrees? I'll make a PR.

**Comment by xu-cheng** [comment]
Fixed by 12201b17c00270cf28d47bfc073e6cb7c8c9799d
